exporter/opentelemetry-exporter-otlp/tests/performance/benchmarks/test_benchmark_trace_exporter.py

```diff
@@ -0,0 +1,74 @@
+# Copyright The OpenTelemetry Authors
+#
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+#     http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+from unittest.mock import patch
+
+from opentelemetry.exporter.otlp.trace_exporter import OTLPSpanExporter
+from opentelemetry.sdk.trace import TracerProvider, sampling
+from opentelemetry.sdk.trace.export import (
+    BatchExportSpanProcessor,
+    SimpleExportSpanProcessor,
+)
+
+
+def get_tracer_with_processor(span_processor_class):
+    simple_span_processor = span_processor_class(OTLPSpanExporter())
+    tracer = TracerProvider(
+        active_span_processor=simple_span_processor,
+        sampler=sampling.DEFAULT_ON,
+    ).get_tracer("pipeline_benchmark_tracer")
+    return tracer
+
+
+class MockTraceServiceStub(object):
+    def __init__(self, channel):
+        self.Export = lambda *args, **kwargs: None
+
+
+@patch(
+    "opentelemetry.exporter.otlp.trace_exporter.OTLPSpanExporter._stub",
+    new=MockTraceServiceStub,
+)
+def test_simple_span_processor(benchmark):
+    tracer = get_tracer_with_processor(SimpleExportSpanProcessor)
+
+    def create_spans_to_be_exported():
+        span = tracer.start_span("benchmarkedSpan",)
+        for i in range(10):
+            span.set_attribute(
+                "benchmarkAttribute_{}".format(i),
+                "benchmarkAttrValue_{}".format(i),
+            )
+        span.end()
+
+    benchmark(create_spans_to_be_exported)
+
+
+@patch(
+    "opentelemetry.exporter.otlp.trace_exporter.OTLPSpanExporter._stub",
+    new=MockTraceServiceStub,
+)
+def test_batch_span_processor(benchmark):
+    tracer = get_tracer_with_processor(BatchExportSpanProcessor)
+
+    def create_spans_to_be_exported():
```

Comment:
If you want to be complete, you can run [force_flush](https://github.com/open-telemetry/opentelemetry-python/blob/master/opentelemetry-sdk/src/opentelemetry/sdk/trace/__init__.py#L99). 

It'll require some calculation, but you may be better off re-writing this benchmark to run a combination of 10k create_spans_to_be_exported, followed by a force_flush. This ensures a consistent benchmark, rather than the batch span processor running when it needs to, which could be different from run to run. 

But it looks like max_queue_size on the batch span exporter will force flushes anyway. (default [2048](https://github.com/open-telemetry/opentelemetry-python/blob/master/opentelemetry-sdk/src/opentelemetry/sdk/trace/export/__init__.py#L126))

Reply:
Thanks for your super helpful comments as always :)

I thought a lot about what you said, and it makes sense to me, but I also think it makes sense to keep these `BatchExportSpanProcessor` tests as is with only 1 span created every time.

I noticed that when I changed it to "create 2048 spans and then export" you could no longer compare the "Processors" because the Batch Processor could only run 5 times (but it exported 2048 * 5 = 10,240 spans) while the Span Processor ran 134 times. Further, I reasoned that if that 2048 * 4th span finished right before the 1 second limit by `pytest-benchmark`, it would run the function again and add another 2048 spans to the benchmark when really the test should have been over by that point (and it is for the `SimpleExportSpanProcessor`).

It's true that the "time-to-export" for the `BatchExportSpanProcessor` won't be consistent, but when I ran it locally it never got so inconsistent that it seemed to produce crazy results. Over my trials, I got the following # of spans completed in 1 second:

Trial 1: 8036
Trial 2: 7282
Trial 3: 7965
Trial 4: 7642
Trial 5: 7871

I'm not sure _when_ the Processor decides to export, but I think what's great about benchmarking 1 single span creation allows us to see how much _better_ it is than the `SingleExportSpanProcessor` here over the same trials:

Trial 1: 138
Trial 2: 134
Trial 3: 134
Trial 4: 129
Trial 5: 139

This means the `BatchExportSpanProcessor` never regresses so badly in its "time-to-export" decisions that it gets slow like the `SimpleExportSpanProcessor` which exports on every span.

Once we have time to include [Self-hosted runners](https://docs.github.com/en/free-pro-team@latest/actions/hosting-your-own-runners/about-self-hosted-runners) I would expect its decisions to get consistent, but even so, these tests will make sure that regressions in the Processor "time-to-export" decisions don't go unnoticed :)

TL;DR I think it's better for both comparisons against other Processors and accuracy for # of spans finished during the 1s benchmark to leave the test as is. This is because even the "Batch" tests should be consistent enough to not have sharp changes unless the algorithm changes.